### PR TITLE
fixed for few major issues with grid

### LINKED ISSUE #3141: Changing Grid RowDefinitions/ColumnDefinitions does not invalidate visual
When RowDefinitions or ColumnDefinitions change (instantiate new Definitions and set DirectProperty on Grid) changes are not shown until something else triggers the control to render/invalidate.

(Tested on 0.9 Preview 2 & 3)

### LINKED ISSUE #3155: Grid throws exception when removing children.
Grid will throws following exception when you remove child in last column or last row ([repo](https://bitbucket.org/igorstefus/grid_problem/src/master/GridProblem/)):

```
Unhandled Exception: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV, Boolean& hasDesiredSizeUChanged) in D:\a\1\s\src\Avalonia.Controls\Grid.cs:line 991
   at Avalonia.Controls.Grid.MeasureOverride(Size constraint) in D:\a\1\s\src\Avalonia.Controls\Grid.cs:line 487
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize) in D:\a\1\s\src\Avalonia.Layout\Layoutable.cs:line 516
   at Avalonia.Layout.Layoutable.Measure(Size availableSize) in D:\a\1\s\src\Avalonia.Layout\Layoutable.cs:line 317
   at Avalonia.Layout.LayoutManager.Measure(ILayoutable control) in D:\a\1\s\src\Avalonia.Layout\LayoutManager.cs:line 177
   at Avalonia.Layout.LayoutManager.ExecuteMeasurePass() in D:\a\1\s\src\Avalonia.Layout\LayoutManager.cs:line 128
   at Avalonia.Layout.LayoutManager.ExecuteLayoutPass() in D:\a\1\s\src\Avalonia.Layout\LayoutManager.cs:line 90
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority) in D:\a\1\s\src\Avalonia.Base\Threading\JobRunner.cs:line 40
   at Avalonia.Win32.Win32Platform.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in D:\a\1\s\src\Windows\Avalonia.Win32\Win32Platform.cs:line 174
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken) in D:\a\1\s\src\Windows\Avalonia.Win32\Win32Platform.cs:line 121
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in D:\a\1\s\src\Avalonia.Base\Threading\Dispatcher.cs:line 65
   at GridProblem.Program.AppMain(Application app, String[] args) in D:\Repos\GridProblem\GridProblem\Program.cs:line 28
   at GridProblem.Program.Main(String[] args) in D:\Repos\GridProblem\GridProblem\Program.cs:line 14
```

### PR DESCRIPTION
**Pull request by donandren**
## What does the pull request do?

- fixes grid when column/row property is changed grid to react properly
- fixes grid when column/row is added/removed grid to react properly
- fixes crash when child is added/removed runtime from grid after initialization

## What is the current behavior?

grid is not affected when column/row properties are changed or when added/removed

## What is the updated/expected behavior with this PR?

fix the major issues with grid above

## How was the solution implemented (if it's not obvious)?

Invalidade Measure is triggered when column/row properties are changed or when added/removed

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?

## Breaking changes

nope
## Fixed issues

Fixed #3141 
Fixed #3155

### CODE REVIEW

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 1]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
             Assert.Equal(new Size(100, 100), grid.Bounds.Size);
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Change_Column_Width_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: Nits:  `Change_Column_Width_Should_Invalidate_Grid` -> `Changing_Column_Width_Should_Invalidate_Grid` 

Other similar tests should have the same correction too

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 2]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+            {
+                if (setUsingAvaloniaProperty)
+                    grid.RowDefinitions[0][RowDefinition.MaxHeightProperty] = 5;
+                else
+                    grid.RowDefinitions[0].MaxHeight = 5;
+            });
+        }
+
+        [Fact]
+        public void Add_Column_Should_Invalidate_Grid()
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: Nit: `Add_` -> `Adding_`

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 3]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
+            {
+                grid.RowDefinitions.Add(new RowDefinition(new GridLength(5)));
+            });
+        }
+
+        [Fact]
+        public void Replace_Columns_Should_Invalidate_Grid()
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: Nit: `Replace_` -> `Replacing_`

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 4]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
+            {
+                grid.ColumnDefinitions.RemoveAt(0);
+            });
+        }
+
+        [Fact]
+        public void Remove_Row_Should_Invalidate_Grid()
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: Nit: `Remove_` -> `Removing_`

File: src/Avalonia.Controls/DefinitionBase.cs [hunk 5]
```diff
@@ -730,5 +734,22 @@ static DefinitionBase()
             SharedSizeGroupProperty.Changed.AddClassHandler<DefinitionBase>(OnSharedSizeGroupPropertyChanged);
             PrivateSharedSizeScopeProperty.Changed.AddClassHandler<DefinitionBase>(OnPrivateSharedSizeScopePropertyChanged);
         }
+
+        /// <summary>
+        /// Marks a property on a definition as affecting the parent grid's measurement.
+        /// </summary>
+        /// <param name="properties">The properties.</param>
+        protected static void AffectsParentMeasure(params AvaloniaProperty[] properties)
```
Comment (jmacato) [inline src/Avalonia.Controls/DefinitionBase.cs]: I thought this was a standard function on any control... Guess this should be considered in 0.10 for a unified approach on this.

**Review by jmacato — CHANGES_REQUESTED** [review]
Great work! Haven't had time to fix these lately... Just some grammar nits, hope you don't mind :)

File: src/Avalonia.Controls/DefinitionBase.cs (hunk 5 above)
Comment (akunchev) [reply src/Avalonia.Controls/DefinitionBase.cs]: there is AffectsParentMeasure already, but it's for controls only, here `DefinitionBase` is not a control so it need custom non control implementation

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 6]
```diff
@@ -1382,6 +1376,182 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
+            {
+                grid.RowDefinitions.RemoveAt(0);
+            });
+        }
+
+        private static void Change_Propery_And_Verify_Measure_Requested(Grid grid, Action change)
```
Comment (MarchingCube) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]: Typo in `Propery` -> `Property`.

**Review by MarchingCube — APPROVED** [review]
LGTM with a minor typo to fix.

File: src/Avalonia.Controls/DefinitionBase.cs (hunk 5 above)
Comment (jmacato) [reply src/Avalonia.Controls/DefinitionBase.cs]: @akunchev good catch!

**Review by jmacato — APPROVED** [review]

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 7]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Column_Width_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 8]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Column_MinWidth_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 9]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Column_MaxWidth_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 10]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Row_Height_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 11]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Row_MinHeight_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 12]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+        }
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void Changing_Row_MaxHeight_Should_Invalidate_Grid(bool setUsingAvaloniaProperty)
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 13]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                    grid.RowDefinitions[0].MaxHeight = 5;
+            });
+        }
+
+        [Fact]
+        public void Adding_Column_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 14]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                grid.ColumnDefinitions.Add(new ColumnDefinition(new GridLength(5)));
+            });
+        }
+
+        [Fact]
+        public void Adding_Row_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 15]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                grid.RowDefinitions.Add(new RowDefinition(new GridLength(5)));
+            });
+        }
+
+        [Fact]
+        public void Replacing_Columns_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 16]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                grid.ColumnDefinitions = ColumnDefinitions.Parse("2*,1*");
+            });
+        }
+
+        [Fact]
+        public void Replacing_Rows_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 17]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                grid.RowDefinitions = RowDefinitions.Parse("2*,1*");
+            });
+        }
+
+        [Fact]
+        public void Removing_Column_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { ColumnDefinitions = ColumnDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 18]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+                grid.ColumnDefinitions.RemoveAt(0);
+            });
+        }
+
+        [Fact]
+        public void Removing_Row_Should_Invalidate_Grid()
+        {
+            var grid = new Grid { RowDefinitions = RowDefinitions.Parse("1*,1*") };
+
+            Change_Propery_And_Verify_Measure_Requested(grid, () =>
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
            Change_Property_And_Verify_Measure_Requested(grid, () =>
```

File: tests/Avalonia.Controls.UnitTests/GridTests.cs [hunk 19]
```diff
@@ -1382,6 +1376,245 @@ public void Correct_Grid_Bounds_When_Child_Control_Has_DesiredSize_Larger_Than_A
...
+            grid.Arrange(new Rect(size));
+
+            Assert.True(grid.IsMeasureValid);
+            Assert.True(grid.IsArrangeValid);
+
+            Assert.Equal(90, grid.Children[0].Bounds.Width);
+            Assert.Equal(10, grid.Children[1].Bounds.Width);
+        }
+
+        private static void Change_Propery_And_Verify_Measure_Requested(Grid grid, Action change)
```
Comment (jmacato) [inline tests/Avalonia.Controls.UnitTests/GridTests.cs]:
```suggestion
        private static void Change_Property_And_Verify_Measure_Requested(Grid grid, Action change)
```

**Review by jmacato — CHANGES_REQUESTED** [review]

**Review by jmacato — APPROVED** [review]

**Pull request merged**